design/open/spark-outputcommitter/committer.md

```diff
@@ -0,0 +1,194 @@
+committer for S3A.  It requires a "staging" filesystem on HDFS and offers
+manual modes for overwriting outputs ("conflict resolution").
+
+These issues remain with the magic and staging committer:
+
+* Not atomic.  Multipart objects appear sequentially.  Finding a "_success"
+  object indicator is still required before a directory tree can be
+  processed.
+* The magic committer requires "magic committer" support in the FileSystem,
+  in order to write `__magic` objects to a _different_ path.  (Current
+  lakeFSFS does not contain this support.)
+* The magic committer is somewhat new:
+  > It’s also not been field tested to the extent of Netflix’s committer;
+  > consider it the least mature of the committers.
+* Documentation is still somewhat lacking.
+
+Other OutputCommitters are not as good, and _rename by copying_.  This is
+both slow and highly non-atomic, as well as making it difficult to recover
+from failed attempts.  The current recommendations are to use the
+partitioned staging committer for overwriting or updating partitioned data
+trees, and the magic committer is most other cases.
+
+## LakeFSOutputCommitter
+
+We propose to leverage the atomic capabilities of lakeFS to write a specific
+OutputCommitter for lakeFSFS.  In the initial version, it will branch out,
+prepare the desired output, and merge back in as part of the HOC commit.[^1]
+Aborting will be done by dropping the branch (or repurposing it if the same
+job ID is requested again).  Regular lakeFS retention can handle dropping
+data objects; we might want to add file patterns to retention to allow
+temporary objects to be dropped rapidly.
+
+[^1]: A merge is a type of commit, of course, so in this model HOC commits
+	*are* lakeFS commits!
+
+# How
+
+## Conflict modes
+
+Spark supports multiple "save modes": Append, Overwrite, ErrorIfExists, and
+Ignore.  These impact conflict resolution.  We will initially support just
+overwrite modes: an entire previous table will be deleted on write.
+
+## Sample flow
+
+* User configures lakeFSFS and configures LakeFSOutputCommitter as the
+  OutputCommitter for `lakefs` protocol Paths.  **Possibly** lakeFSFS will
+  set up this OutputCommitter by default.
+
+  Successive steps are controlled by Spark/Hadoop to output, and correspond
+  to the Hadoop OutputCommitter protocol.
+* **Setup (job/task TBD)**: Create a new branch for this job/task.  Its name
+  is predictable from the job ID and/or task ID, so can be easily found
+  again.  In "overwrite" mode, immediately delete the entire subtree of the
+  intended output path, and possibly delete branches of previous tasks.
+  This handles cases where the names of the objects written by the output
+  format change, for instance because of repartitioning to a smaller (or
+  different) number of partitions or because of nondeterminism in the names
+  of the objects.
+* **Write objects**: Everything is written to its correct path on the branch.
```

Comment:
Does this mean that output names are still dynamic but due to the delete step the HOC does, we get to overwrite the previous output written to write the destination path?

Reply:
Sort-of.  We don't control path names, so they could change during an overwrite.  OutputCommitters don't really change final pathnames.  The change here is that objects are only ever written to their correct pathnames.  Again: it is much easier to see and understand what did what and see the source of each object.